sse2neon.h

```diff
@@ -8358,6 +8359,495 @@ FORCE_INLINE int _mm_testz_si128(__m128i a, __m128i b)
 
 /* SSE4.2 */
 
+const static uint16_t _sse2neon_cmpestr_mask16b[8] ALIGN_STRUCT(16) = {
+    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80,
+};
+const static uint8_t _sse2neon_cmpestr_mask8b[16] ALIGN_STRUCT(16) = {
+    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80,
+    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80,
+};
+
+/* specify the source data format */
+#define _SIDD_UBYTE_OPS 0x00 /* unsigned 8-bit characters */
+#define _SIDD_UWORD_OPS 0x01 /* unsigned 16-bit characters */
+#define _SIDD_SBYTE_OPS 0x02 /* signed 8-bit characters */
+#define _SIDD_SWORD_OPS 0x03 /* signed 16-bit characters */
+
+/* specify the comparison operation */
+#define _SIDD_CMP_EQUAL_ANY 0x00     /* compare equal any: strchr */
+#define _SIDD_CMP_RANGES 0x04        /* compare ranges */
+#define _SIDD_CMP_EQUAL_EACH 0x08    /* compare equal each: strcmp */
+#define _SIDD_CMP_EQUAL_ORDERED 0x0C /* compare equal ordered */
+
+/* specify the polarity */
+#define _SIDD_POSITIVE_POLARITY 0x00
+#define _SIDD_MASKED_POSITIVE_POLARITY 0x20
+#define _SIDD_NEGATIVE_POLARITY 0x10 /* negate results */
+#define _SIDD_MASKED_NEGATIVE_POLARITY \
+    0x30 /* negate results only before end of string */
+
+/* specify the output selection in _mm_cmpXstri */
+#define _SIDD_LEAST_SIGNIFICANT 0x00
+#define _SIDD_MOST_SIGNIFICANT 0x40
+
+/* specify the output selection in _mm_cmpXstrm */
+#define _SIDD_BIT_MASK 0x00
+#define _SIDD_UNIT_MASK 0x40
+
+/* Pattern Matching for C macros.
+ * https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms
+ */
+
+/* catenate */
+#define PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
+
+#define IIF(c) PRIMITIVE_CAT(IIF_, c)
+/* run the 2nd parameter */
+#define IIF_0(t, ...) __VA_ARGS__
+/* run the 1st parameter */
+#define IIF_1(t, ...) t
+
+#define EQ_u8(i) EQ(u8, i)
+
+#define EQ_u16(i) EQ(u16, i)
+
+#define EQ(type, i)                                                            \
+    mtx[i] = vreinterpretq_m128i_##type(vceqq_##type(                          \
+        vdupq_n_##type(vgetq_lane_##type(vreinterpretq_##type##_m128i(b), i)), \
+        vreinterpretq_##type##_m128i(a)));
+
+#define RNG_u8(i) RNG(u8, i)
+
+#define RNG_s8(i) RNG(s8, i)
+
+#define RNG_u16(i) RNG(u16, i)
+
+#define RNG_s16(i) RNG(s16, i)
+
+#define RNG(type, i) \
+    vec_b[i] =       \
+        vdupq_n_##type(vgetq_lane_##type(vreinterpretq_##type##_m128i(b), i));
+
+#define x8(f, i) \
+    f(i + 0) f(i + 1) f(i + 2) f(i + 3) f(i + 4) f(i + 5) f(i + 6) f(i + 7)
+
+#define x16(f, i) x8(f, i) x8(f, i + 8)
+
+#define PCMPSTR_RNG_U_8x16(a, b, mtx)                                     \
+    do {                                                                  \
+        uint8x16_t vec_b[16];                                             \
+        __m128i mask = vreinterpretq_m128i_u16(vdupq_n_u16(0xff));        \
+        x16(RNG_u8, 0) for (int i = 0; i < 16; i++)                       \
```

Comment:
If you pass types such as `u8` to macro `x16`, it is feasible to avoid defining macros like `RNG_u8`.